File: SHM_15_Classes_with_XPointer_Reference.R
# This time the class uses a reference to the XPtr and does not own it itself.
# 
# More information about XPtr:
# https://dirk.eddelbuettel.com/code/rcpp/html/classRcpp_1_1XPtr.html
# ----------------------------------

Rcpp::sourceCpp("SHM_15_Classes_with_XPointer_Reference.cpp")

# create external pointer obj to our matrix
xptr_mat <- create_XPtr_for_R_obj(matrix(rnorm(12), 3, 4))
# check out the obj
print(xptr_mat)
xptr::xptr_address(xptr_mat)


# use external pointer as input to class obj creation where the
# the class uses the exact same data pointer to the obj
external_mat_admin <- new(External_Matrix_Administration, xptr_mat)
external_mat_admin$print_Matrix()
external_mat_admin$add_42_to_Matrix()
external_mat_admin$print_Matrix()
external_mat_admin$return_Matrix()
external_mat_admin$add_number_to_Matrix(42.42)
external_mat_admin$print_Matrix()
external_mat_admin$return_Matrix()
external_mat_admin$add_number_to_Matrix(42L)
external_mat_admin$print_Matrix()
external_mat_admin$return_Matrix()
# this time the return pointer is exactly the same as the input pointer
external_mat_admin$return_pointer_to_Data() 

# object no longer needed: delete it, release pointers and memory
rm(external_mat_admin)

# check out the obj
xptr::xptr_address(xptr_mat)
print(xptr_mat)

# release memory
xptr::xptr_clear(xptr_mat)
xptr::xptr_address(xptr_mat)
print(xptr_mat)
# the memory is now released, delete pointer obj
rm(xptr_mat)

# NOTE: the External_Matrix_Administration objs is still running(!) if we delete
#       the external pointer obj and call the class obj again. 
#       But if we perform a garbage collection (gc()) the memory address is lost.
#       Invoking the call from the class towards the data results into a 
#       crash.

File: SHM_15_Classes_with_XPointer_Reference.cpp
// [[Rcpp::depends(RcppArmadillo)]]
#include <iostream>
#include <RcppArmadillo.h>


class External_Matrix_Administration {
  
public:
  
  // ctor
  // --------------------
  External_Matrix_Administration (Rcpp::XPtr<arma::mat>&& xptr_A) {
    Rcpp::Rcout << xptr_A << std::endl;
    // In this case the class uses the same pointer! Be aware, that using the
    // move just counts for the C++ backend. In R the object is still
    // present under its original name.
    this->armaMat = std::move(xptr_A);
  };
  
  // dtor
  // --------------------
  ~External_Matrix_Administration() {
    armaMat = nullptr;
    delete armaMat;
  };
  
  // Member method
  // --------------------
  void print_Matrix() const{
    armaMat->print();
    Rcpp::Rcout << std::endl;
  }
  
  void add_42_to_Matrix() {
    *armaMat += 42;
  }
  
  void add_number_to_Matrix(const double number) {
    *armaMat += number;
  }

  arma::mat return_Matrix() const{
    return *armaMat;
  }
  
  // memory address of data, same as with xptr_A
  Rcpp::XPtr<arma::mat> return_pointer_to_Data() const {
    return Rcpp::XPtr<arma::mat>(armaMat);
  }
  
private:
  // Attribute
  // --------------------
  arma::mat* armaMat{nullptr};
};


Rcpp::XPtr<arma::mat> create_XPtr_for_R_obj(arma::mat& A) {
  arma::mat* armaMat = new arma::mat(A.memptr(), A.n_rows, A.n_cols, true, false);
  return Rcpp::XPtr<arma::mat> (armaMat);
}


RCPP_MODULE(mod_External_Matrix_Admin) {
  using namespace Rcpp;

  function("create_XPtr_for_R_obj",
           &create_XPtr_for_R_obj,
           List::create(_["mat"]), "Create external pointer of R matrix obj");
  

  Rcpp::class_<External_Matrix_Administration>("External_Matrix_Administration")
   .constructor<Rcpp::XPtr<arma::mat>>("Creates new class obj.")
   .method("print_Matrix", &External_Matrix_Administration::print_Matrix, "Prints matrix.")
   .method("add_42_to_Matrix", &External_Matrix_Administration::add_42_to_Matrix, "Add 42 to matrix.")
   .method("return_Matrix", &External_Matrix_Administration::return_Matrix, "Return matrix to R.")
   .method("add_number_to_Matrix", &External_Matrix_Administration::add_number_to_Matrix, "Add number to Matrix.")
   .method("return_pointer_to_Data", &External_Matrix_Administration::return_pointer_to_Data, "Return pointer to data.")
  ;
  
}